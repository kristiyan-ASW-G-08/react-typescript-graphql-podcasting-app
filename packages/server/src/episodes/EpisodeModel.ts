import mongoose, { Schema } from 'mongoose';
import EpisodeType from '@customTypes/EpisodeType';
import uniqueValidator from 'mongoose-unique-validator';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';

const EpisodeSchema = new Schema<EpisodeType>({
  title: { type: String, maxlength: 50, minlength: 1 },
  description: { type: String, required: true },
  audioFile: { type: String, required: true },
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  podcast: {
    type: Schema.Types.ObjectId,
    ref: 'Podcast',
    required: true,
  },
  date: {
    type: Date,
    default: Date.now,
  },
});

// // @ts-ignore
// EpisodeSchema.plugin(uniqueValidator);
// // @ts-ignore
// EpisodeSchema.post('save', duplicationErrorHandler);
// // @ts-ignore
// EpisodeSchema.post('update', duplicationErrorHandler);

export default mongoose.model<EpisodeType>('Episode', EpisodeSchema);
