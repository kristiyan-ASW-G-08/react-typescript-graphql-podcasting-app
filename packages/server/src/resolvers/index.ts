import userMutations from 'src/users/usersMutations';
import podcastMutations from '@src/podcasts/podcastsMutations';
import podcastQueries from '@src/podcasts/podcastQueries';
import episodeMutations from '@src/episodes/episodeMutations';

const resolvers = {
  Query: {
    ...podcastQueries,
  },
  Mutation: { ...userMutations, ...podcastMutations, ...episodeMutations },
};

export default resolvers;
