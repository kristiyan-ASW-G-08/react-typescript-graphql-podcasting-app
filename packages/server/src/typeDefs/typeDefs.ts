import { gql } from 'apollo-server';

const typeDefs = gql`
  scalar Upload

  type User {
    username: String!
    email: String!
    _id: ID!
  }

  type Query {
    getUser: User!
  }

  type File {
    filename: String!
    mimetype: String!
    encoding: String!
  }
  type UserLoginPayload {
    username: String!
    email: String!
    userId: String!
    token: String!
  }
  type Podcast {
    title: String!
    website: String!
    cover: String!
    user: User!
    _id: ID!
  }

  type Episode {
    title: String!
    description: String!
    audioFile: String!
    date: String!
    user: User!
    podcast: Podcast!
    _id: ID!
  }
  type Episodes {
    episodes: [Episode]!
  }
  type Query {
    podcasts(query: String): [Podcast]!
    getPodcast(podcastId: ID!): Podcast!
    getEpisode(episodeId: ID!): Episode!
    getEpisodesByPodcast(podcastId: ID!): Episodes!
  }
  type Mutation {
    createUserMutation(
      username: String!
      email: String!
      password: String!
      confirmPassword: String!
    ): UserLoginPayload!

    loginMutation(email: String!, password: String!): UserLoginPayload!

    createPodcastMutation(
      title: String!
      website: String!
      cover: Upload!
    ): Podcast!

    updatePodcastMutation(
      title: String!
      website: String!
      cover: Upload
      _id: ID!
    ): Podcast!

    createEpisodeMutation(
      title: String!
      description: String!
      audioFile: Upload!
      podcast: ID!
    ): Episode!
  }
`;

export default typeDefs;
