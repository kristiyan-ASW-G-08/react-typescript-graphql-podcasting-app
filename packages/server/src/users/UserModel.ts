import mongoose, { Schema } from 'mongoose';
import UserType from '@src/types/UserType';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: { type: String, minlength: 1, maxlength: 50, unique: true },
  email: { type: String, unique: true },
  password: { type: String, minlength: 8 },
});

UserSchema.plugin(uniqueValidator);
// @ts-ignore
UserSchema.post('save', duplicationErrorHandler);
// @ts-ignore
UserSchema.post('update', duplicationErrorHandler);

export default mongoose.model<UserType>('User', UserSchema);
