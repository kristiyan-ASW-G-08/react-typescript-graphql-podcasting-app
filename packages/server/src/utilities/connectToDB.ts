import mongoose from 'mongoose';
import logger from '@customUtilities/logger';

const connectToDB = async (mongoURI: string): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
  } catch (err) {
    logger.error(`MongoDB Connection Error: ${err}`);
  }
};
export default connectToDB;
