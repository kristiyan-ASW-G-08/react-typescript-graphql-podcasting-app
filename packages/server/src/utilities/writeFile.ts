import fs from 'fs';
import path from 'path';

const writeFile = (
  filename: string,
  stream: NodeJS.ReadWriteStream,
): string => {
  // @ts-ignore
  const filePath = path.join(
    path.basename(path.dirname(filename)),
    `/images/${new Date().toISOString()}-${filename}`,
  );
  stream.pipe(fs.createWriteStream(filePath));
  return filePath;
};

export default writeFile;
