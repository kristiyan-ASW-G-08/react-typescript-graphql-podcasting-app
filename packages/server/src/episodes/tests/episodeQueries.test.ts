import mongoose from 'mongoose';
import { ApolloServer } from 'apollo-server-express';
import { sign } from 'jsonwebtoken';
import connectToDB from '@customUtilities/connectToDB';
import UserModel from '@src/users/UserModel';
import EpisodeModel from '@src/episodes/EpisodeModel';
// @ts-ignore
import typeDefs from '@typeDefs/typeDefs';
import { Readable } from 'stream';
// @ts-ignore
import { Upload } from 'graphql-upload/public';
import writeFile from '@customUtilities/writeFile';
import resolvers from '../../resolvers/index';

jest.mock('@customUtilities/writeFile');

(writeFile as jest.Mock).mockReturnValue('/images/some_file.svg');
describe('Resolvers', () => {
  const { MONGO_USER, MONGO_PASSWORD, MONGO_DATABASE } = process.env;
  const mongoURI = `mongodb+srv://${MONGO_USER}:${MONGO_PASSWORD}@cluster0.ol9wi.mongodb.net/${MONGO_DATABASE}?retryWrites=true`;
  beforeAll(async () => {
    await mongoose.disconnect();
    await connectToDB(mongoURI);
    await EpisodeModel.deleteMany({}).exec();
  });

  afterEach(async () => {
    await EpisodeModel.deleteMany({}).exec();
  });
  afterAll(async () => {
    await mongoose.disconnect();
  });
  const values = {
    title: 'asd',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    podcast: mongoose.Types.ObjectId().toString(),
    user: mongoose.Types.ObjectId().toString(),
    audioFile: 'someAudiofile',
  };
  describe('Queries', () => {
    describe('getEpisode', () => {
      it('should return a episode successfully', async () => {
        const episode = await new EpisodeModel({ ...values }).save();
        const episodeId = episode._id.toString();
        expect.assertions(2);

        const query = `
          query getEpisode($episodeId:ID!) {
            getEpisode(episodeId:$episodeId) {
              _id
            }
          }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
          variables: {
            episodeId,
          },
        });
        expect(result.errors).toBeUndefined();
        expect(result.data?.getEpisode._id).toBeTruthy();
      });

      it('should throw an error when episode is not found', async () => {
        expect.assertions(1);

        const query = `
          mutation getEpisode( $episodeId:ID!) {
            getEpisode(episodeId:$episodeId) {
              _id
            }
          }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });
        const result = await testServer.executeOperation({
          query,
          variables: {
            episodeId: mongoose.Types.ObjectId(),
          },
        });
        expect(result.errors).toMatchSnapshot();
      });
    });
  });

  describe('getEpisodesByPodcast', () => {
    it('should return a episodes successfully', async () => {
      const episode = await new EpisodeModel({ ...values }).save();
      expect.assertions(2);

      const query = `
        query getEpisodesByPodcast($podcastId:ID!) {
          getEpisodesByPodcast(podcastId:$podcastId) {
            episodes{
              _id

            }
          }
        }
      `;

      const testServer = new ApolloServer({
        typeDefs,
        resolvers,
      });

      const result = await testServer.executeOperation({
        query,
        variables: {
          podcastId: episode.podcast.toString(),
        },
      });
      expect(result.errors).toBeUndefined();
      expect(result.data?.getEpisodesByPodcast.episodes).toBeTruthy();
    });
  });
});
