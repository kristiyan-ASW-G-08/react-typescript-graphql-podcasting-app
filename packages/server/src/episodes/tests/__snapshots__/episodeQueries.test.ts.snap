// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers Queries getEpisode should throw an error when episode is not found 1`] = `
Array [
  [ValidationError: Cannot query field "getEpisode" on type "Mutation".],
]
`;
